Lib/subprocess.py

```diff
@@ -1480,7 +1480,26 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             if shell:
                 startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
                 startupinfo.wShowWindow = _winapi.SW_HIDE
-                comspec = os.environ.get("COMSPEC", "cmd.exe")
+                # gh-101283: without a fully-qualified path, before Windows
+                # checks the system directories, it first looks in the
+                # application directory, and also the current directory if
+                # NeedCurrentDirectoryForExePathW(ExeName) is true, so try
+                # to avoid executing unqualified "cmd.exe".
+                comspec = os.environ.get('ComSpec')
+                if not comspec:
+                    system_drive = os.environ.get('SystemDrive') or 'C:'
+                    system_root = os.environ.get('SystemRoot') or os.path.join(
+                        system_drive, 'Windows')
+                    comspec = os.path.join(system_root, 'System32', 'cmd.exe')
+                    if not os.path.isfile(comspec):
```

Comment:
Agreed for the `main` branch. I think it is worth keeping as a fallback in the release branch backports to avoid breaking someones working code.

I'm not the right person to understand if such a non \Windows\System32 based environment without the environment variables set actually ever exists in the real world.

Reply:
Unfortunately, it's really easy to spawn a subprocess and specify a partial environment that doesn't inherit anything. Typically the child process won't even be able to start under these circumstances, but depending on how someone fixes it can affect how likely this is (e.g. add variables until it works? Might not add `COMSPEC`).

Reply:
This would significantly change the behavior. Currently we let `CreateProcessW()` search for "cmd.exe" if `ComSpec` isn't defined. The API doesn't use the `SystemRoot` or `SystemDrive` environment variables in `RtlGetExePath()`, the function that builds the executable search path. It builds the search path as follows:

* application directory, i.e. the dynamic environment variable `__APPDIR__` 
* current directory, i.e. ("."), if `NeedCurrentDirectoryForExePathW(name)` is true
* system directory, i.e. `GetSystemDirectoryW()`); e.g. "C:\\Windows\\System32"
* legacy WIN16 system directory (no API query); e.g. "C:\\Windows\\System"
* Windows directory, i.e. `GetSystemWindowsDirectoryW()`; e.g. "C:\\Windows"
* `PATH`

I'd prefer to fall back on using the system directory from `GetSystemDirectoryW()`. If we use that, then I'm comfortable with letting `_winapi.CreateProcess()` fail if "cmd.exe" isn't found there. I'm less comfortable with giving up if the `SystemRoot` or `SystemDrive` environment variable is wrong. I don't think that's the same as `ComSpec` being wrong, which is a shell path that we've been explicitly told to use. If `ComSpec` is wrong, failing is the best response.

Reply:
I'm also okay with failing without `ComSpec` and not using `SystemRoot`.

One other mid-ground that might save us adding a native method right away could be to read `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment@ComSpec` directly, which at least requires an administrator to modify, rather than just the parent process.

Reply:
FYI, for `os.system(command)`, if `ComSpec` is defined and not empty, the `_wsystem()` call  tries `_wspawnv(_P_WAIT, comspec, {comspec, slash_c, command, NULL})`. The latter checks for the existence of `comspec` relative to the current working directory. If the filename has no extension, it successively checks `comspec` plus ".com", ".exe", ".bat", and ".cmd". It executes the first name that exists. If the `_wspawnv()` call fails, `_wsystem()` falls back on `_wspawnvp(_P_WAIT, L"cmd.exe", {L"cmd.exe", slash_c, command, NULL})`. This successively tries `_wspawnv()` for the current directory and each of the `PATH` directories. The C runtime thus has an unsafe fallback. That's yet another reason to implement `os.system()` with `subprocess.Popen`, as was already implemented for `os.popen()`.

Reply:
The "ComSpec" value in the session manager's "Environment" registry key is a good default[^1]. That way we'll use the system shell that was configured by an administrator (e.g. maybe it's [TCC/LE](https://jpsoft.com/products/tcc-le.html)), rather than defaulting to "cmd.exe".

The default installed value of `ComSpec` is the `REG_EXPAND_SZ` value "%SystemRoot%\\system32\\cmd.exe". We can expand the system `ComSpec` value with `winreg.ExpandEnvironmentStrings()`. This API function doesn't have any special handling for known system variables such as `SystemRoot`, but it does have the advantage that it doesn't expand an environment variable if the value is an empty string. `ntpath.expandvars()`, in contrast, will expand an environment variable to an empty string.

If we're implementing strict behavior, we could fail immediately if the default shell path isn't a fully-qualified path with a drive and root. But leave it up to `CreateProcessW()` to check for existence.

[^1]: Though I'd rather we supported [`CreateEnvironmentBlock()`](https://learn.microsoft.com/en-us/windows/win32/api/userenv/nf-userenv-createenvironmentblock) in `os`. For example, `os.create_environ(system_only=False, inherit=False)`. The result would be an `os._Environ` mapping with keys that are case insensitive. If [issue 7310](https://github.com/python/cpython/issues/73010) is resolved, the keys will also be case preserving.

Reply:
Here's an example that moves the work into a method. This example disallows using a relative path for the shell, unless it's from the `ComSpec` environment variable. If it has to fall back on "%SystemRoot%\\System32\\cmd.exe", it will fail if `SystemRoot` is missing or empty.

```python
def _isabspath(self, s):
    # ntpath.isabs() returns true for a rooted path such as "\System32".
    # A rooted path depends on the drive of the current working
    # directory, or, if it's the target path of a symlink, on the drive
    # of the opened path of the symlink. This method implements the
    # correct behavior. 
    s = s[:3].replace('/', '\\')
    return s[:2] == '\\\\' or (s[1:] == ':\\' and s[:1] != '\\')

def _get_shell_path(self):
    # gh-101283: avoid using a relative path for the shell, unless the
    # the current ComSpec environment variable is a relative path.
    # Fall back on the system's configured ComSpec value, or else on
    # "%SystemRoot%\System32\cmd.exe".
    comspec = os.environ.get('ComSpec')
    if comspec:
        return comspec
    import winreg
    try:
        hkey = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE,
                r'System\CurrentControlSet\Control'
                r'\Session Manager\Environment')
        comspec, vt = winreg.QueryValueEx(hkey, 'ComSpec')
        if not comspec or vt not in (
                winreg.REG_EXPAND_SZ, winreg.REG_SZ):
            raise ValueError('bad ComSpec registry value: '
                             f'{comspec}!r')
        if vt == winreg.REG_EXPAND_SZ:
            comspec = winreg.ExpandEnvironmentStrings(comspec)
    except (OSError, ValueError):
        comspec = winreg.ExpandEnvironmentStrings(
                    r'%SystemRoot%\System32\cmd.exe')
    if not self._isabspath(comspec):
        raise ValueError('default shell path is not absolute: '
                         f'{comspec!r}')
    return comspec
```

Usage:

```python
                shell_path = self._get_shell_path()
                if self._isabspath(shell_path):
                    executable = shell_path
                args = f'"{shell_path}" /c "{args}"'
```

Note that `shell_path` in the command line has to be quoted, in case the path contains spaces. That's true even if we use `executable` because the shell has to parse its command line.